File: midterm/load_balancer.cpp
#include <iostream>
#include <vector>
#include <string>
#include <random>
#include <cstring>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <curl/curl.h>

// List of Reddit server addresses
std::vector<std::string> redditServers = {"https://www.reddit.com/r/subreddit1", "https://www.reddit.com/r/subreddit2", "https://www.reddit.com/r/subreddit3"};

// Load balancer server address and port
std::string lbAddress = "192.168.1.200";
int lbPort = 8080;

// Callback function for libcurl
static size_t writeCallback(char *contents, size_t size, size_t nmemb, void *userp) {
    ((std::string*)userp)->append(contents, size * nmemb);
    return size * nmemb;
}

int main() {
    // Initialize libcurl
    curl_global_init(CURL_GLOBAL_DEFAULT);

    // Create a socket for the load balancer
    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        std::cerr << "Error creating socket" << std::endl;
        return 1;
    }

    // Bind the socket to the load balancer address and port
    struct sockaddr_in serv_addr;
    memset(&serv_addr, 0, sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(lbAddress.c_str());
    serv_addr.sin_port = htons(lbPort);
    if (bind(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
        std::cerr << "Error binding socket" << std::endl;
        return 1;
    }

    // Listen for incoming connections
    if (listen(sockfd, 5) < 0) {
        std::cerr << "Error listening on socket" << std::endl;
        return 1;
    }

    std::cout << "Load balancer listening on " << lbAddress << ":" << lbPort << std::endl;

    while (true) {
        // Wait for a connection
        int clientfd = accept(sockfd, NULL, NULL);
        if (clientFd < 0) {
            std::cerr << "Error accepting connection" << std::endl;
            continue;
        }

        std::cout << "Received connection" << std::endl;

        // Choose a Reddit server
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, redditServers.size() - 1);
        int idx = dis(gen);
        std::string redditServer = redditServers[idx];
        std::cout << "Forwarding request to Reddit server: " << redditServer << std::endl;

        // Forward the request to the Reddit server
        CURL* curl = curl_easy_init();
        if (curl) {
            std::string response;
            curl_easy_setopt(curl, CURLOPT_URL, redditServer.c_str());
            curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
            curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
            CURLcode res = curl_easy_perform(curl);
            if (res != CURLE_OK) {
                std::cerr << "Error forwarding request to Reddit server: " << curl_easy_strerror(res) << std::endl;
                send(clientFd, "Error: Could not connect to Reddit server.", 40, 0);
            } else {
                send(clientFd, response.c_str(), response.length(), 0);
            }
            curl_easy_cleanup(curl);
        }

        // Close the client connection
        close(clientFd);
    }
    

    // Clean up
    close(sockfd);
    curl_global_cleanup();
    return 0;
}

File: midterm/load_balancer.py
import socket
import random
import requests

# List of Reddit server addresses
reddit_servers = ['https://www.reddit.com/r/Music/', 'https://www.reddit.com/r/musictheory/', 'https://www.reddit.com/r/red_velvet/']

# Load balancer server address and port
lb_address = '127.0.0.1'
lb_port = 8080

# Create a socket for the load balancer
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

sock.bind((lb_address, lb_port))
sock.listen(5)

# Round-robin (most basic) load balancing algorithm
next_server_index = 0

def handle_request_round_robin(next_server_index):
    reddit_server = reddit_servers[next_server_index]
    next_server_index = (next_server_index + 1) % len(reddit_servers)
    return reddit_server, next_server_index

# Least connections load balancing algorithm
def handle_request_least_connections(next_server_index):
    reddit_server = min(reddit_servers, key=lambda server: server.get_connection_count())
    return reddit_server, next_server_index


print(f'Load balancer listening on {lb_address}:{lb_port}')

while True:
    # Wait for a connection
    conn, addr = sock.accept()
    print(f'Received connection from {addr}')

    # Choose a Reddit server
    reddit_server, next_server_index = handle_request_round_robin(next_server_index)


    print(f'Forwarding request to Reddit server: {reddit_server}')

    # Forward the request to the Reddit server
    try:
        response = requests.get(reddit_server)
        response.raise_for_status()
    except requests.exceptions.RequestException as e:
        print(f'Error forwarding request to Reddit server: {e}')
        conn.sendall(b'Error: Could not connect to Reddit server.')
        conn.close()
        continue

    # Send the response back to the client
    conn.sendall(response.content)
    conn.close()